### PR TITLE
expression: rewrite builtin function: HOUR, MINUTE, SECOND, MICROSECOND

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -462,30 +462,27 @@ func (c *hourFunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinHourSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpDuration)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen, bf.tp.Decimal = 2, 0
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: mysql> SELECT HOUR('272:59:59');
        -> 272

File: expression/builtin_time.go [hunk 2]
```diff
@@ -564,30 +555,27 @@ func (c *microSecondFunctionClass) getFunction(args []Expression, ctx context.Co
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinMicroSecondSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpDuration)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen, bf.tp.Decimal = 21, 0
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: MICROSECOND(expr)

Returns the microseconds from the time or datetime expression expr as a number in the range from 0 to 999999.

File: expression/integration_test.go [hunk 3]
```diff
@@ -893,6 +893,28 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 	tk.MustExec("INSERT INTO t VALUES (0), (20030101010160), (20030101016001), (20030101240101), (20030132010101), (20031301010101), (20031200000000), (20030000000000);")
 	result = tk.MustQuery("SELECT CAST(ix AS SIGNED) FROM t;")
 	result.Check(testkit.Rows("0", "0", "0", "0", "0", "0", "0", "0"))
+
+	//for hour
+	result = tk.MustQuery(`SELECT hour("12:13:14.123456"), hour("12:13:14.000010"), hour("272:59:55"), hour(null);`)
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: add test cases like 
``` sql
select hour("2017-12-12 12:12:12");
```

File: expression/builtin_time.go (hunk 1 above)
Comment (zz-jason) [inline expression/builtin_time.go]: that's a problem, in mysql:
```sql
MySQL > SELECT HOUR('272:59:59');
Field   1:  `HOUR('272:59:59')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONGLONG
Collation:  binary (63)
Length:     2
Max_length: 3
Decimals:   0
Flags:      BINARY NUM


+-------------------+
| HOUR('272:59:59') |
+-------------------+
|               272 |
+-------------------+
1 row in set (0.00 sec)
```

File: expression/builtin_time.go (hunk 2 above)
Comment (zz-jason) [inline expression/builtin_time.go]: in mysql:

```sql
MySQL > SELECT microsecond('272:59:59');
Field   1:  `microsecond('272:59:59')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONGLONG
Collation:  binary (63)
Length:     21
Max_length: 1
Decimals:   0
Flags:      BINARY NUM


+--------------------------+
| microsecond('272:59:59') |
+--------------------------+
|                        0 |
+--------------------------+
1 row in set (0.00 sec)
```

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by zz-jason** [comment]
@jackysp PTAL

**Pull request merged**